Structs with no fields or all-ZST fields are ZSTs (#2262)
rust-lang/reference: src/type-layout.md

Reviewer: Now that structs guarantee the same, the exception for unit tuples is no longer needed.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/61c35c046e32955fae60107dd0b9c58de788aadc..08c156565eb9fc1741d8bd97fd95116d9b66f4f8)*

Reviewer 2: I think technically we don't guarantee that empty `repr(Rust)` structs are 1-aligned (yet), only that they are 0-sized.

Also, I think for clarity it'd be better to keep this section regardless, but maybe rephrase it to not say it's an exception, just a note or something.

Be aware that this guarantee does not imply that the fields have distinct addresses: [zero-sized types] may have the same address as other fields in the same struct.

r[layout.repr.rust.struct-zst]

> For [structs] with no fields, or where all fields are [zero-sized], it is further guaranteed that the structs are themselves [zero-sized].

Reviewer 3: ```suggestion
For [structs] with no fields, or where all fields are [zero sized], it is further guaranteed that the structs are themselves [zero sized].
```

When used as a phrasal adjective, let's hyphenate this.  In predicate position, let's not (I know the Reference isn't yet consistent about this today — we're moving in that direction).

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/86635e30bf861a038dc197d7e16fd09e7e514e7a..fa61cbdf2a146d3f7a16cbe0d10648f81e3626dc)*

> For [structs] with no fields, or where all fields are [zero-sized], it is further guaranteed that the structs are themselves [zero-sized].

Reviewer 3: ```suggestion
For [structs] with no fields or where all fields are [zero-sized], it is further guaranteed that the structs are themselves [zero-sized].
```

(Remove the comma there.)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/86635e30bf861a038dc197d7e16fd09e7e514e7a..fa61cbdf2a146d3f7a16cbe0d10648f81e3626dc)*

r[layout.repr.rust.enum-struct-like-zst]

> For [enums] (without a primitive representation specified) with a single struct-like variant with no fields or where all fields are [zero-sized], the enum itself is [zero-sized].

Reviewer 4: Seems bizarre to guarantee this for struct-like variants but not unit-like variants or tuple-like variants.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/4fd61400e123abc713b8b0b3b1f52bc2bfb56888..5011dcdb90f3b84599ace73a4ee0e05c3de7d286)*

> For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant where the struct-like thing has no fields or where all of the fields are [zero-sized], the enum itself is [zero-sized].

Reviewer 3: ```suggestion
For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant and where the struct-like thing has no fields or where all of the fields are [zero sized], the enum itself is [zero sized].
```

(This came directly from my FCP text.  Let's add an "and" to help disambiguate the scope of the "where".

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/86635e30bf861a038dc197d7e16fd09e7e514e7a..fa61cbdf2a146d3f7a16cbe0d10648f81e3626dc)*

> For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant where the struct-like thing has no fields or where all of the fields are [zero-sized], the enum itself is [zero-sized].

Reviewer 3: ```suggestion
For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant where the struct-like thing has no fields or where all of the fields are [zero-sized], the enums themselves are [zero sized].
```

Let's match the plurality pattern of the struct case above.  (Again, I'm correcting my earlier verbiage here.)

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/86635e30bf861a038dc197d7e16fd09e7e514e7a..fa61cbdf2a146d3f7a16cbe0d10648f81e3626dc)*

> For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant and where the struct-like thing has no fields or where all of the fields are [zero sized], the enums themselves are [zero sized].

Reviewer 3: Let's add links to cross-reference each of these, here and in the glossary.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/61c35c046e32955fae60107dd0b9c58de788aadc..034e3cf6cbc086cd9a3f5a302d627f8a91f34a77)*

> For [enums] (without a primitive representation specified) with a single field-struct-like variant, a single unit-struct-like variant, or a single tuple-struct-like variant and where the struct-like thing has no fields or where all of the fields are [zero sized], the enums themselves are [zero sized].

Reviewer 3: Let's add links to cross-reference each of these, here and in the glossary.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2262/61c35c046e32955fae60107dd0b9c58de788aadc..034e3cf6cbc086cd9a3f5a302d627f8a91f34a77)*
